### PR TITLE
Auto-configuration of OpenSC PKCS#11

### PR DESCRIPTION
**Pull request by frankmorgner**
In this PR I've implemented `pkcs11-register` to auto-configure Mozilla programs and Chrome for the use of (onepin-)opensc-pkcs11 module. Since a) I didn't find a way to do this globally and b) I wanted to avoid the dependency to the NSS libraries, I've chosen to simply modify `pkcs11.txt` of the NSS db directory when the user is logging in. (This requires the directory to be in *new* db format, which is the default since NSS 3.35.)

This auto-configuration is tested for Firefox on macOS, Linux and Windows. After installing MSI/PKG and restarting Firefox, the module will be configured and ready to for use without further user interaction.

Some minor fixes for the build process are also included, see the commit messages.

##### Checklist

- [ ] Documentation is added or updated
- [x] New files have a LGPL 2.1 license statement
- [x] PKCS#11 module is tested

### CODE REVIEW

**Comment by Jakuje** [comment]
FYI in Fedora, we did this using `p11-kit-proxy`, which is loaded by every NSS application by default:

https://fedoraproject.org/wiki/Changes/NSSLoadP11KitModules

**Comment by frankmorgner** [comment]
I also thought about p11-kit-proxy.so/dylib for longer. Theoretically, I could add an exception to not add OpenSC to pkcs11.txt, when this PKCS#11 provider is already registered (which boils down to directly checking via the NSS API, which I decided to avoid). However, I think, the setup you're using for fedora is not very widespread. For example, on Debian, p11-kit comes completely unconfigured (and I guess on Ubuntu it will be similar). [`modutil` requires an additional confirmation if the user wants to load an additional module next to p11-kit-proxy.so](https://github.com/nss-dev/nss/blob/924a14158d2371947bf4877974647b8f4f32b425/cmd/modutil/modutil.c#L863-L864), but still adds the module. Anyway, I think it's fair to simply pass `--enable-autostart-items` on well configured p11-kit boxes. Is it possible to detect such a configuration?

File: MacOSX/Distribution.xml.in [hunk 1]
```diff
@@ -20,4 +20,7 @@ https://developer.apple.com/library/mac/documentation/DeveloperTools/Reference/D
     <choice id="tokend" title="OpenSC tokend">
         <pkg-ref id="org.opensc-project.tokend">OpenSC-tokend.pkg</pkg-ref>
     </choice>
+    <choice id="startup" title="OpenSC startup icons">
```
Comment (metsma) [inline MacOSX/Distribution.xml.in]: choices-outline missing reference to startup

File: MacOSX/scripts/postinstall [hunk 2]
```diff
@@ -15,8 +15,18 @@ else
 	mv /Library/OpenSC/etc/opensc.conf.orig /Library/OpenSC/etc/opensc.conf
 	md5 -r /Library/OpenSC/etc/opensc.conf  > /Library/OpenSC/etc/opensc.conf.md5
 fi
+
 for f in /Library/OpenSC/bin/*
 do
 	ln -sf $f /usr/local/bin
 done
+
+for f in /Library/LaunchAgents/pkcs11-register.plist /Library/LaunchAgents/pkcs11-register.plist
```
Comment (metsma) [inline MacOSX/scripts/postinstall]: Twice the same in loop ? "/Library/LaunchAgents/pkcs11-register.plist"

Comment (frankmorgner) [reply MacOSX/scripts/postinstall]: copy/paste error. thanks, fixed.

**Comment by Jakuje** [comment]
Actually, the p11-kit is pulled in the NSS by the crypto-policies at this moment using this configuration file (not sure whether it is symlinked from somewhere else):

  /etc/crypto-policies/back-ends/nss.config

I do not mind excluding this part as part of Fedora/RHEL spec file, but having this covered right here would be more elegant.

**Comment by frankmorgner** [comment]
> I do not mind excluding this part as part of Fedora/RHEL spec file, but having this covered right here would be more elegant.

Sure, but I don't know how to check this in a platform (linux) independent manner... In configure.ac, I could check `modutil -list` for p11-kit-proxy.so, but that doesn't cover all the possible database directories that are relevant...

**Comment by frankmorgner** [comment]
@Jakuje any ideas on how to systematically avoid problems with p11-kit?

**Comment by Jakuje** [comment]
@frankmorgner I am not sure what you are asking now and what problems are you talking about now. You mean the prompt?

**Comment by frankmorgner** [comment]
My question was whether we could add anything to configure.ac or the c-code to avoid double registering the OpenSC module by accident if this is done by both, p11-kit and pkcs11-register. If `--disable-autostart-items` is already enough, then this PR can be merged, I think.

**Comment by Jakuje** [comment]
I do not think there is something that could be used during configure/build time. In case of Fedora, we should get the same results in builders and installed systems, but generally it does not have to be the same if the user configured the p11-kit manually. I think this should be runtime check. But I do not think you will be able to figure this out reliably without using the nss-tools or NSS API.

In any case, adding @ueno, the NSS and p11-kit developer, whether he has some comments to the way how the NSS DB is manipulated here and whether there is a simple way to detect NSS has p11-kit loaded either in configure time or in runtime in platform-independent way.

The simplest approximation is probably to check in configure, whether there is p11-kit installed and if so, do `--disable-autostart-items`.

**Pull request merged**